[ConstraintSystem] Don't check `-disable-availability-checking`

xedin opened the pull request:
Checking for `-disable-availability-checking` in
`ConstraintSystem::isDeclUnavailable` caused a regression with
obsolete/introduced checking. Let's rely on
`DeclAttributes::isUnavailable` and `TypeChecker::isDeclAvailable`
to do the right thing instead.

Resolves: rdar://problem/60898369

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Don't check `-disable-availability-checking`

Checking for `-disable-availability-checking` in
`ConstraintSystem::isDeclUnavailable` caused a regression with
obsolete/introduced checking. Let's rely on
`DeclAttributes::isUnavailable` and `TypeChecker::isDeclAvailable`
to do the right thing instead.

Resolves: rdar://problem/60898369

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test macOS platform

DougGregor commented on test/Constraints/rdar60898369.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+func test(bookmark: Data, relativeURL: URL?, resolutionOptions: URL.BookmarkResolutionOptions) throws {
```
Could we eliminate the Foundation and `objc_interop` dependencies here? This should reproduce with Swift 4.2 @availables on all platforms.

xedin replied:
That's what I thought but I couldn't reduce this unfortunately, here is what I tried:

```
struct S {
  @available(swift, obsoleted: 4.2)
  init(_: Int) throws {}

  @available(swift, introduced: 4.2)
  init?(_: Int) throws {}
}

_ = try S(42)
```

But it type-checks correctly even without this change. I'm not as proficient in all of the special cases related to availability checking to figure out what exactly I had to do...

theblixguy replied:
```swift
struct S {
  @available(swift, obsoleted: 4.2)
  init(foo: Int) throws {}

  @available(swift, introduced: 4.2)
  init?(foo: Int) throws {}
}

_ = try S(foo: 42)
```

This gives me `ambiguous use of 'init(foo:)'`. I guess the label is important?

xedin replied:
Interesting, I guess so...

xedin replied:
Thank you, @theblixguy! I have updated the test-case.

xedin commented:
@swift-ci please smoke test

xedin committed:
[ConstraintSystem] Switch availability check to use correct API

Use `TypeChecker::checkDeclarationAvailability` instead of `isDeclAvailable`,
that is a proper API endpoint which does flag checking before calling
`isDeclAvailable` internally.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test Windows platform
